File: 509.fibonacci-number.cpp
/*
 * @lc app=leetcode id=509 lang=cpp
 *
 * [509] Fibonacci Number
 */

#include <bits/stdc++.h>
#include <cassert>

using namespace std;

// @lc code=start
class Solution {
 public:
  int fib(int n) {
    if (n == 0) {
      return 0;
    } else if (n == 1) {
      return 1;
    }
    return fib(n - 1) + fib(n - 2);
  }
};
// @lc code=end

int
main () {
  Solution s;
  assert(s.fib(0) == 0);
  assert(s.fib(1) == 1);
  assert(s.fib(2) == 1);
  assert(s.fib(3) == 2);
  assert(s.fib(4) == 3);
  assert(s.fib(5) == 5);
  return 0;
}

File: 509.fibonacci-number.rs
/*
 * @lc app=leetcode id=509 lang=rust
 *
 * [509] Fibonacci Number
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn fib(n: i32) -> i32 {
        if n == 0 {
            return 0;
        } else if n == 1 {
            return 1;
        }
        let mut ppnum = 0;
        let mut pnum = 1;

        let mut ret = 1;
        for _i in 0..(n - 1) {
            ret = pnum + ppnum;
            ppnum = pnum;
            pnum = ret;
        }
        ret
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(Solution::fib(2), 1);
}

#[test]
fn test2() {
    assert_eq!(Solution::fib(3), 2);
}

#[test]
fn test3() {
    assert_eq!(Solution::fib(4), 3);
}
